### DISCUSSION TITLE
vba-articles/why-do-people-use-vba

### DISCUSSION THREAD

**Discussion by giscus** [sancarn/vba-articles#7] [_giscus]
# vba-articles/why-do-people-use-vba

Introduction Recently, I watched a video by YouTuber ThePrimeTime which details a dev’s frustrations with business culture. Prime is an ex-entrepreneur who currently works in software development at Netflix. His views in this video have been criticised for being jaded by FAANG business cultures he has worked in. I personally don’t feel this way. Although there is some truth to the developer’s (mataroa’s) article, I think it misses the root causes of many issues raised. I have mixed feelings about the article in question. Some areas I agreed with and others I disagreed with. However I did want to address one particular concern / statement: I work on a platform that cost my organization an eye-watering sum of money to produce, over the span of two years, and the engineers responsible for it elected to use spreadsheets to control the infrastructure, so we now have a spreadsheet with 400 separate worksheets that powers but one part of this whole shambling mess. I’m speculating here, but I’d imagine that the business/SMEs (Subject Matter Experts) are using VBA to some capacity to control their 400 worksheet collection. So this begs the question… Why do people use VBA? In order to answer this question, we must first look at another question - who actually uses VBA in the first place? In 2021 I ran a poll on /r/vba where I asked redditors why they code in VBA. From these data, we can clearly see that the majority of people who use VBA do so mainly because they have no other choice. Many organisations run their entire business processes with Excel, and when a little bit of automation is required VBA is usually #1 on the list. The versatility of VBA In the business I currently work for, in the engineering division, we have access to a variety of technologies (automation platforms): OnPrem - PowerShell (No access to Install-Module) OnPrem - Excel (VBA / OfficeJS (limited access) / OfficeScripts / PowerQuery) OnPrem - PowerBI Desktop OnPrem - SAP Analysis for Office OnCloud - Power Platform (PowerApps, Power BI, PowerAutomate (non-premium only)) SandboxedServer - ArcGIS (ArcPy) SandboxedServer - MapInfo (MapBasic) SandboxedServer - InfoWorks ICM (Ruby) SandboxedCloud - ArcGIS Online We also have a number of databases controlled by IT: D1. OnPrem - Geospatial database D2. OnCloud - Geospatial mirror D3. OnPrem - SAP database D4. OnCloud - SAP BW4HANA partial mirror D5. OnPrem - Telemetry platform D6. OnPrem - Sharepoint D7. OnCloud - Sharepoint Online D8. OnCloud - EDM Telemetry platform D9. OnCloud - Large mirror database D10. OnPrem - LotusNotes database D11. OnPrem - IBM BPM database D12. OnPrem - File System D13. OnPrem - Hydraulic Model Information D1-D13 databases are summarised in the table below listing what types of data are stored in which systems, the importance of the data stored in each database, and whether the database is essentially a replica of OnPrem information: Data Type D1 D2 D3 D4 D5 D6 D7 D8 D9 D10 D11 D12 D13 Customer Issues X X X X         X         Asset Data X X X X   X X   X     X X Telemetry Data         X     X           Risk Data X X         X     X X   X Financial Data     X X               X   Misc Data X X       X X         X X Note: D6 would be C tier if it weren’t for the fact we continue to store a business critical spreadsheet on Sharepoint OnPrem for compatibility reasons. See the control of VBA for details. And the data’s importance / on cloud replication: Data Type D1 D2 D3 D4 D5 D6 D7 D8 D9 D10 D11 D12 D13 Data Importance S S S S S S/C A B S S S S A Replica of N/A D1 N/A ~D3 N/A N/A N/A N/A ~D4 N/A N/A N/A N/A Note: Online replicas are de facto replicas in terms of data’s importance, although the reason we need to connect to them is diminished Finally, let’s examine how our Automation Platforms link to our Data Platforms. Links shown in the diagram are where the automation platform can access the data from the various data platforms/sources: Note: Some of the links from VBA to OnCloud services are based on my attempts alone. There is no doubt in my mind that VBA can interface with SAP BW4HANA and our other cloud services, I just haven’t figured out the authentication requirements and protocols yet Here’s where you might start to see an issue. Looks like the only automation platforms which can connect to all the data sources we need is VBA and Powershell. Power BI Desktop has been introduced in our business but doesn’t hit all the platforms which VBA does, and even if it did Power BI cannot be used for process automation where-as VBA can, so what’s the point making the switch? Users who do use Power BI to target these other datasets usually generate CSVs of this other data and store these in cloud sharepoint system, but what generates those CSVs? VBA. Now, we’d love to use a higher level language in our organisation to handle this business automation. However, every request for a high level language to be installed across the team/business e.g. Python / Ruby / Node / Rust etc. has been rejected by CyberSecurity in favour of technologies like PowerAutomate, PowerApps which as you can see above barely touch any of the data we need. It is supposedly “Against the technology strategic vision of the company” to allow “end-users” access to high level programming languages. Now even if the data access was there in our business, PowerPlatform would still be insufficient to perform the majority of our processes because the algorithms required are so complex that a PowerAutomate solutions would become infuriating to maintain and incomprehensible to even IT folks (e.g. See projection algorithms). Ultimately the stand-out technologies for us are Powershell v3 (doesn’t even support class syntax and cannot install modules), and VBA, purely from a versatility standpoint. As a result of this ‘monopoly’ on technology I and others have spent hundreds of hours building open source VBA libraries which augment VBA promoting it to a reasonable language by modern standards. The maintenance guarantee of VBA D10 and D11 above are intimately linked. In 2000s many of our systems were built on top of IBM Lotus Notes databases. In 2019 Lotus Notes was acquired by HCL, and since then longevity of support has been wavering. Support will officially die in June 2024. As a result, since 2019, technology teams have been trying to migrate many of our systems to new technologies. The business spent an eye watering amount of money developing a system using IBM Business Process manager to supercede one of these Lotus Notes databases. The anticipation was that D11 would be backfilled with all the data from D10, once fully built, and D10 would be archived. It’s now 2023: We are 8 months away from official support dying. Technology teams have thrown away their support contact for IBM BPM. There is no replacement in sight for both IBM BPM and Lotus Notes databases. IBM BPM solution is poorly maintained IBM BPM solution has numerous issues and doesn’t function as needed Solution has been shoehorned into IBM BPM, despite the platform not being fit for purpose i.e. while IBM BPM does come with a REST API, this REST API is borderline useless to Technology teams and SMEs Some REST calls use javascript encoded as strings Others require html embedded in json embedded in xml Database tables aren’t queried by name but by GUID. There’s no documentation of which GUID relates to which table/process. The data from D10 was never actually transferred to D11, meaning the business is now using 2 systems instead of 1. D11 data model doesn’t really support the data in D10 either. Technology teams appear to be ignoring waning support contracts. This could be due to lack of resources, finances and/or priority. SME’s use these tools on a daily basis, and ultimately it is SME’s who determine changes that need to be made to the system. If SMEs use VBA, they can control and maintain the system as needed. They have a maintenance guarantee, something that should be said for IT systems too, but can’t be. The control of VBA In a recent project we are building a new all encompassing IT system to supercede a business critical spreadsheet. This would ultimately demote D6 to C tier importance. The spec for this system was initially simple - Give us a NodeJS server with a MySQL database. Use React for the UI. Give admins/SMEs (subject matter experts) access to the codebase with access to git for code control. IT and SMEs will collaborate to build the system. Technology teams demanded that admins/SMEs will not have access to the code. Technology teams demanded that FrontEnd be built in Microsoft PowerApps, to comply with “Strategic Vision”. Technology teams demanded that BackEnd be built in Microsoft Azure Pipelines, to comply with “Strategic Vision”. Unfortunately, as an admin/SME with more development knowledge than many people in technology, these demands do not sit well with me: Technology teams do not understand work in the teams thus do not understand business logic and calculations Thus devs writing business logic is error prone. Technology teams has frequently abandoned bespoke technology projects, leaving no resource to maintain and improve the system. Collaboration with SMEs will ensure that at least 1 team maintains resource to maintain system. SMEs need to ensure that they have confidence in what is produced. How, without observing that the code doesn’t work for all edge cases? Unit tests? Perhaps, but without seeing the code how can we verify these unit tests exist? And are ran frequently*. SMEs improve and maintain the existing legacy system, and have unparalleled knowledge of how systems interact. Less knowledge can be shared in upskilling Technology teams where it is required. SMEs need to ensure all data is transferred and represented correctly in new system. SMEs unable to do this without backend access. Ultimately, as long as code stays in VBA it is controlled by the SMEs and the business. Technology teams rarely relinquish control to business teams. SMEs can ensure that software is developed properly in a modular fashion and doesn’t end up as a cluster of barely working technologies loosely linked together. The UX of VBA This is a smaller one, because you can always make familiar feeling tools in any language, but there is something to be said about the fact that most engineers use spreadsheets in their day job, and how VBA is embedded within these spreadsheets and can improve on UX. Giving someone a foreign tool in a familiar environment, is often much more powerful to that user, than a foreign tool in a foreign environment. Conclusion In conclusion, yes, we (and many others in businesses) do choose to use spreadsheets (and VBA) for many tasks within our organisations, there are many reasons for this, including: Poor alternatives provided by IT due to security concerns. Poor connectivity of alternatives to source systems, usually because they are still WIP. Faults in IT strategy which don’t account for certain use-cases. Unwillingness to collaborate with SMEs due to security and maintenance concerns. Lack of training for users/managers/SMEs in alternative systems. Users/SMEs wanting some level of control over the business logic in these systems. It’s the only viable technology which is available to everyone, as it’s part of Office. This does not mean that we are at all blind to VBA’s weaknesses though: Why is VBA most dreaded? What is wrong with VBA? There’s no doubt in my mind that there are some elements of truth to mataroa’s article. Sometimes management is poor, but more often than not I believe most people in organisations are trying to do the right thing, and are doing whatever they can with the tools that are available to them.

https://sancarn.github.io/vba-articles/why-do-people-use-vba.html

**Comment by arh5451** [comment]
Hi. I love your article. I am a Data/AI/Strategy/BI Consultant. You hit the head on the nail with the IT department bottle necking processes. From my experience it continues to be the case that IT controlled "business systems" i.e. data systems, is the cause of most of the pain in organizations. 

I see this as a big reason why companies adopt cloud today. Not because of "scalability" or "efficiency", but to get the data out of IT hands and into those of the business.

**Reply by sancarn** [reply]
> I see this as a big reason why companies adopt cloud today. Not because of "scalability" or "efficiency", but to get the data out of IT hands and into those of the business.

I don't think it's about getting data out of the hands of IT personally, I think cloud adoption is mainly for accessibility of data and ease of maintenance. Without really good engineers managing interfaces between microservices the scalability claims are dubious. However as discussed in the article, anecdotally, most of our data is on-prem, and most of our automation tools are on-prem too. So cloud migration isn't that useful...

I do really think that IT architects/strategy planners need to ensure the fundamental tools for automation of cloud data are in place (and imo ensure low-code tools aren't the only game in town) and available prior to migration of data to cloud services. Only then will cloud migration be useful to business users.

**Reply by arh5451** [reply]
100% agree. I like to simplify things. Perhaps this is an over-simplification; however, no matter what tool you use requires good management, design, and decision making.

I just know from experience if you have a good decision-maker in charge of cloud platform it can be really powerful because they are only beholden to budget and not bureaucracy.

**Comment by hiAndrewQuinn** [comment]
I was in a quite similar position for the first year of my career as well - the environments I found myself in were heavily gated enough that I wound up getting really good at PowerShell, despite always being more of a Linux guy at heart.

People really don't think enough about the strange market dynamics of things like PowerShell and VBA. The demand for people who can work with these technologies is huge, precisely because they're so ubiquitous in non-tech shops (and even tech shops with strict rules in place as to who gets to program with what where) -- the *supply* of dev talent willing to work on it is miniscule.

**Comment by davejnewman** [comment]
As a VBA developer for over 20 years I see the issue as IT thinking they own the data instead of the business owning it. In my experience I have worked with formal IT methodologies and found that the business is hampered by most of them as the IT department is slow to respond.
Why?
Any operational business needs reporting. The reporting built into large applications such as SAP is limited to data screens and Business Objects (which was actually usable until SAP bought it and removed VBA...)To expect a quick report from IT is ridiculous so business learns to use VBA. 
The above being true, the solution is for the business to retake the ownership from IT and make it do it's job.
Data protection by limiting access of non business users.
Maintenance of data systems including hardware, networking and database management.
IMHO, get IT back in it's box and out of the way.

**Reply by adamwdotnet** [reply]
I have seen both ends of this spectrum, and I can assure you that its not a great situation when either party is not involved at all. If IT steps back and gives free reign, you often end up with a total mess, the business user/s leave the business and IT are looked to to fix the problem, or you end up with so many data problems that they are constantly blaming the system for producing bad or inconsistent results. On the flip side, if IT retain too much control, youre absolutely right in that they become a total bottleneck and its difficult for the business to get what they need. A balance is needed, and this is often difficult to achieve.

**Reply by KallunWillock** [reply]
Like adamwdotnet, I've seen both ends of the spectrum, and I agree - things can go wildly wrong if there isn't some kind of balance between the two.

I've worked for an employer where it took 11 months to have pre-approved software installled on my work laptop - a solid 5 months after I actually needed it. Fortunately, I had access to VBA, and was able to reproduce the requested software in time to complete the project. On the other end of the spectrum, I shudder to think what would happen if my colleagues could just copy and paste their way into some semblance of a solution with [insert programming language here] being freely available. So... it's a spectrum.

**Comment by munrobasher** [comment]
I ended up here after spending another fun (no really) day or so coding a pretty big data import for a client in VBA. I've been developing software and working on business systems for many many years and I'm very aware of the limitations of VBA. But I've spent so long developing VBA code, that I've got my own styles, helper classes etc. Sure, there are more elegant languages but you just can't compete with the immediacy of it all - which can also be it's downfall. 

My client is highly proficient in Excel with a smattering of knowledge of VBA. I'm sure there is a more elegant solution but this is a small company and two days of my time would pale into insignificance with moving to a new platform.

I am surprised though that Microsoft hasn't implemented another programming language in Office Microsoft 365 apps as BASIC is really not the way to go. I spotted the article on Python in Excel but that seems to be limited to using Python in cells. Not sure you can write entire modules and classes. I would imagine somebody has written a series of Excel Python classes as wrappers around Excel COM API? A big task in itself.

**Reply by sancarn** [reply]
> I am surprised though that Microsoft hasn't implemented another programming language in Office Microsoft 365 apps as BASIC is really not the way to go.

There is [OfficeJS](https://learn.microsoft.com/en-us/office/dev/add-ins/develop/understanding-the-javascript-api-for-office) and [OfficeScripts](https://learn.microsoft.com/en-us/office/dev/scripts/overview/excel). They could make OfficeScripts into a viable replacement for VBA, but [they aren't interested](https://www.reddit.com/r/IAmA/comments/m22uc6/comment/gqh4fcn/?utm_source=reddit&utm_medium=web2x&context=3).

> I spotted the article on Python in Excel but that seems to be limited to using Python in cells.

Python in Excel, is a little like how jupyter notebooks works. You can evaluate cross-cells and cross-sheets too. But ultimately it's a glorified UDF interface.

**Comment by davejnewman** [comment]
I keep seeing references to VBA having limitations in Office products. Since VBA is built into Office the interface is the quickest method especially if you are building forms.
Yes, there are faster methods to manage importing or getting data into SQL server or Excel but those can be called from within VBA.
I have seen many developers try to use C# to replace applications that were created in Excel or Access. In every case the limitations of the .net interface make the end product less capable.
As to using Python in Excel, as far as
I can tell, it is limited to O365 online and while modules are possible they may not work in desktop. That means the VBA equivalent will still be needed for desktop. I really do not get Microsoft's need to get rid of billions of lines of VBA just to appease short sighted IT managers.

Thanks,
Dave Newman
***@***.***
________________________________

**Reply by sancarn** [reply]
> I have seen many developers try to use C# to replace applications that were created in Excel or Access. In every case the limitations of the .net interface make the end product less capable.

It's strange, I know that's not true. C# and other modern languages are far more limitless than VBA. But I do think coding in VBA breeds a certain type of individual. And those kind of individuals are just difficult to find in any language (or expensive). My opinion is that they look at the employee's salary and search for a developer who's of a similar salary bracket, and usually won't find anyone of the correct level, because the employee was vastly underpaid. And ultimately they hire a significantly less capable developer to replace an existing system, and as a result you end up with a poor application structure and thus less capable product.

Python, currently, is a formulae replacement in Excel. Nothing more. You can't do any actual automation with Python. You can automate to a degree with OfficeJS, but not to the degree that most people use VBA today.

**Reply by munrobasher** [reply]
> Python, currently, is a formualae replacement in Excel. Nothing more. You can't do any actual automation with Python. You can automate to a degree with OfficeJS, but not to the degree that most people use VBA today.

I suspected as much. I'm not that proficient at Python. I sometimes force myself to write a one-off PowerShell script in Python just to keep if reasonably fresh.

**Comment by davejnewman** [comment]
Just to note that because VBA is capable of forms changes that cannot be done in C#. Try adding and deleting fields on a form at runtime. You can even put a form in design mode to make changes.
Agreed that much of this is really bad design but when business wants a firm that changes dynamically they will not settle with less. There is no reason that VBA cannot be held to today's development standards. I know, I did VBA for over 20 years.

Thanks,
Dave Newman
***@***.***
________________________________

**Reply by sancarn** [reply]
> VBA is capable of forms changes that cannot be done in C#. Try adding and deleting fields on a form at runtime. 

Hi Dave, you can actually do this in `C#`, not really sure why your teams couldn't but again I'd put it down to lack of expertise. Perhaps they were using some framework which gave them no control? But both of the main frameworks - WPF and WinForms - both support dynamic controls at both runtime and design time.

> Agreed that much of this is really bad design 

I also wouldn't say it's bad design. Sometimes it's inevitable that this is what you need to solve a problem. Ofc it's something to avoid if possible, but sometimes it is required.

> You can even put a form in design mode to make changes

If you mean to say you can put it into design mode at runtime, yes this is a very unique feature of VBA which most languages have no replacement for! At least, not without significant effort / library.

> There is no reason that VBA cannot be held to today's development standards

I discussed that at some lengths in [Why is VBA most dreaded?](https://sancarn.github.io/vba-articles/why-is-vba-most-dreaded.html)

**Comment by davejnewman** [comment]
Sancarn, Thanks  I tried it a few times and didn't get far but my C# is not the best.  The C# team couldn't figure it out at all.

Thanks,
Dave Newman
***@***.***
________________________________

**Reply by sancarn** [reply]
Hi Dave,

As an example:

## `MainWindow.xaml`

```xaml
<Window x:Class="DynamicWPFAppTest.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:local="clr-namespace:DynamicWPFAppTest"
        mc:Ignorable="d"
        Title="MainWindow" Height="450" Width="800">
    <ScrollViewer>
        <StackPanel x:Name="MyStackPanel" Margin="10">
            <!-- Dynamic content will be added here -->
        </StackPanel>
    </ScrollViewer>
</Window>
```

## `MainWindow.xaml.cs`

```cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace DynamicWPFAppTest
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();
            PopulateForm(new YourObject());
        }

        private void PopulateForm(object obj)
        {
            var properties = obj.GetType().GetProperties();
            foreach (var property in properties)
            {
                string labelContent = property.Name; // Using property name as label, modify as needed
                string fieldValue = property.GetValue(obj)?.ToString() ?? "";

                Label label = new Label();
                label.Content = labelContent;
                MyStackPanel.Children.Add(label); 

                TextBox textBox = new TextBox();
                textBox.Text = fieldValue;
                textBox.TextChanged += (sender, args) => property.SetValue(obj, textBox.Text);
                MyStackPanel.Children.Add(textBox);
            }
        }
    }

    // Define your object here
    public class YourObject
    {
        public string Property1 { get; set; }
        public string Property2 { get; set; }
        // Add other properties as needed
    }
}
```

`PopulateForm` function above will:

1. Take the object supplied to it (can be any object)
2. Identify every property/key name of that object 
3. Dynamically populate a form with multiple fields and labels, one for each property of the object
4. Dynamically set up `TextChanged` listeners which set the values of the property when the text field is changed.

Note because `C#` has lambdas this is already much easier than it is in VBA. Listeners in C# take functions as parameters, allowing us to define a short lambda, where as in VBA you'd usually have to have multiple classes to route the TextChange events to one single function. Here it's fairly easy to read and understand. 

Not sure why your C# team was unable to figure it out, as it really took me no more than 10 minutes to find and test in VIsualStudio as a C# dev with very little experience. So I'd imagine either it's a skill issue, or there's some decision they made before to make their lives easier which is now impacting their flexibility. E.G. they might be using JSON to build the UI or something, or they may be passing the whole form lifecycle creation to a library, and now they'd have to rebuild the whole form and all of its components. I suppose in VBA we don't have these issues because there are no libraries (for better or for worse).

**Comment by davejnewman** [comment]
Thanks for the sample code. I am guessing at runtime I can change the list and force a redraw of the page.
I think the problem with the C# team is they are generally web developers and don't "get" desktop at all.

A comment on the VBA and calling functions by name, you can use Application.run "nameoffunction", "listoffunctionparameters")
There is no return value so you have to standardize the called functions to update a module level global.
I used this in a PowerPoint app so the clients could add slides to a folder and put parameters into a .ini file (yes, I am that old)

Thanks,
Dave Newman
***@***.***
________________________________

**Reply by sancarn** [reply]
> at runtime I can change the list and force a redraw of the page

Indeed, though it is unusual you have to force a redraw in higher level languages like C#, but yeah maybe.

> I think the problem with the C# team is they are generally web developers and don't "get" desktop at all.

This code is identical to using JQuery. If they are web developers they should know how to use JavaScript and should likely not be using C# at all... Would be interesting to hear their opinions on this, but I imagine either they feel it's too much effort or they're ignorant.

> A comment on the VBA and calling functions by name, you can use Application.run ... There is no return value

There is a return value actually! You just need to use brackets as common with VBA. If it's a sub it'll return empty. I use this mechanism in [stdCallback](https://github.com/sancarn/stdVBA/blob/master/src/stdCallback.cls), and to be fair I have built [`stdLambda`](https://github.com/sancarn/stdVBA/blob/master/src/stdLambda.cls) which makes building dynamic forms very similar to C# too 👍 :

```vb
Public Function AlterObject(ByVal obj As Object) As Object
  Dim ctrls As Collection: Set ctrls = New Collection
  Dim oForm As frTransformer: Set oForm = New frTransformer
  Dim index As Long: index = -2
  Dim prop
  For Each prop In stdCOM.Create(obj).Properties
    index = index + 2
    ctrls.add stdUIElement.CreateFromType(oForm.controls, uiLabel, prop & "_label", prop, fTop:=index * 20)
    ctrls.add stdUIElement.CreateFromType(oForm.controls, uiTextBox, prop & "_field", , CallByName(obj, prop, VbGet), _
      stdLambda.Create("if $3 = EUIElementEvent.uiElementEventKeyUp then let $1." & prop & " = $2.value end").Bind(obj), _ 
      fTop:=(index + 1) * 20)
  Next
  oForm.Show
  Set AlterObject = obj
End Function
```

**Reply by munrobasher** [reply]
`Dim ctrls As Collection: Set ctrls = New Collection`

Not surprised to see others adopt this style to get around **Dim ctrls As New Collection** :-)

**Reply by sancarn** [reply]
@munrobasher To be fair in this case it wouldn't matter either way 😅 (modal form), but I do find consistency good practice 😁

**Comment by davejnewman** [comment]
Nice code. That would have done well in an Access app I did for truck sales incentive payment forecasting. Each truck model had different characteristics so the columns on the form had to rebuild when the vehicle was changed. The incentive programs changed too so most of the parent and child forms were rebuilt as a model and sales program were selected from drop lists and date ranges. The sql to the back end had to change in sync. The client loved it and it was a blast to build.

Thanks,
Dave Newman
***@***.***
________________________________

**Comment by davejnewman** [comment]
> Dim ctrls As Collection: Set ctrls = New Collection
> Not surprised to see others adopt this style to get around Dim ctrls As New Collection :-)



Generally I prefer creation of Objects at the last moment. The Dim is free and sets the scope. The Set should be as late as possible and remember to Set Object = Nothing

Thanks,
Dave Newman
***@***.***
________________________________

**Reply by sancarn** [reply]
Hi Dave,

FYI responses to emails don't respond in chain/thread! So some of your messages are lacking context on the site itself 😅 This is a github bug:

https://github.com/orgs/community/discussions/3287

If possible, replying directly on github / the site itself is best :) I'm trying to edit your comments accordingly to provide context.

**Reply by davejnewman** [reply]
Sancarn, thanks for letting me know. I will now reply using the link. As you can tell I am new to github. I avoided it for years when I was working and did not have time for it. Now retired but still programming and I find it is better than reddit.

**Reply by sancarn** [reply]
No worries at all. Certainly more of a professional feel on here 👍

**Reply by munrobasher** [reply]
> Generally I prefer creation of Objects at the last moment. 

Agreed - I usually define it on the line before it's needed. Don't like massive DIM statements at the head of a subroutine/function.

>Set Object = Nothing

I wrap almost everything up in a class and most of them have a Cleanup sub to do this. Another area where it would have been easy to add destructor as well as constructor. I find clean up less of an issue in Excel VBA apps compared to VB5.

I've also developed wrappers around some of the common Excel objects like app, workbook, worksheet to implemented additional helper functions. Once again, the lack of object inheritance is a lack in VBA. You can achieve the same aims but it's not as clean.

So most of my bigger Excel apps have the following classes: ExcelApp, ExcelWorkbook, ExcelWorksheet, MyApp and Emerald. All the global module contains is:

Public MyApp As New MyApp
Public ExcelApp As New ExcelApp
Public Emerald As New Emerald

Plus the main entry points visible with Alt-F8 macro picker. The advantage of wrapping functions that would have traditionally been global functions is that they are invisible to the macro picker.

Emerald is my own name for a series of generic VBA routines that I use across Excel, Word, PowerPoint and Outlook. Although used far more in Excel.
